scripts/validate_response.py

```diff
@@ -0,0 +1,102 @@
+"""Response validation using pre-defined question/answer pair."""
+
+import argparse
+import json
+import sys
+
+import requests
+from llama_index.embeddings.huggingface import HuggingFaceEmbedding
+from scipy.spatial.distance import cosine, euclidean
+
+
+# TODO: Add more question/answer pair
+def args_parser(args):
+    """Arguments parser."""
+    parser = argparse.ArgumentParser(description="Response validation module.")
+    parser.add_argument(
+        "-s",
+        "--scenario",
+        choices=["with_rag", "without_rag"],
+        default="with_rag",
+        help="Scenario for which responses will be evaluated.",
+    )
+    parser.add_argument(
+        "-m",
+        "--model",
+        choices=["gpt", "granite"],
+        default="gpt",
+        help="Model for which responses will be evaluated.",
+    )
+    parser.add_argument(
+        "-t",
+        "--threshold",
+        type=lambda val: float(val),
+        default=0.25,
+        help="Threshold value to be used for similarity score.",
+    )
+    return parser.parse_args(args)
+
+
+def main():
+    """Validate llm response."""
+    args = args_parser(sys.argv[1:])
+    print(f"Arguments passed: {args}")
+
+    with open("./scripts/question_answer_pair.json") as qna_f:
+        qa_pairs = json.load(qna_f)
+        qa_pairs = qa_pairs.get(args.model, {}).get(args.scenario, [])
+
+    embedding_model = HuggingFaceEmbedding("BAAI/bge-base-en")
+
+    queries_with_unexpected_response = []
+
+    for qa in qa_pairs:
+        question = qa["question"]
+        answer = qa["answer"]
+
+        response = requests.post(
+            # API question validator is disabled.
+            "http://localhost:8080/v1/query",
+            json={"query": question},
+            timeout=90,
+        ).json()["response"]
+
+        if args.scenario == "without_rag":
+            response = response.replace(
+                "The following response was generated without access to reference content:\n\n",
+                "",
+            )
+
+        res_vec = embedding_model.get_text_embedding(response)
+        ans_vec = embedding_model.get_text_embedding(answer)
+
+        # Distance score
+        cos_score = cosine(res_vec, ans_vec)
+        euc_score = euclidean(res_vec, ans_vec)
+
+        # Naive length consideration with reduced weightage.
+        len_res, len_ans = len(response), len(answer)
+        len_score = (abs(len_res - len_ans) / (len_res + len_ans)) * 0.1
+
+        score = len_score + (cos_score + euc_score) / 2
+        # TODO: Consider both contextual/non-contextual embedding.
+
+        print(
+            f"Question: {question}\n"
+            f"cos_score: {cos_score}, euc_score: {euc_score}, len_score: {len_score}, "
+            f"final_score: {score}"
+        )
```

Comment:
later we might problably export to CSV or any other format, to be able to see + compare previous results. WDYT?